### PR TITLE
Fix L1-L2 message order

### LINKED ISSUE #586: Message order is not guaranteed (L1 -> L2)
**Describe the bug (observed vs expected behavior)**

In the current implementation of the messages from L1 to L2, the logs are fetched for a range of block, but a `HashMap` is used to actually store the logs for each block.

Then, we iterate on this `HashMap`, but as the order is not guaranteed, we may have some blocks processed in the wrong order.

https://github.com/0xSpaceShard/starknet-devnet-rs/blob/fd1e208a85ad9755942970ee538c653e010d874b/crates/starknet-devnet-core/src/messaging/ethereum.rs#L249-L253
https://github.com/0xSpaceShard/starknet-devnet-rs/blob/fd1e208a85ad9755942970ee538c653e010d874b/crates/starknet-devnet-core/src/messaging/ethereum.rs#L166

A fix can be using a `IndexMap` or returning a `Vec` instead of a `HashMap`. But the latter as more implications, as the log is converted into `MessageToL2` (which doesn't contain the block number), the block should be also returned.

**Not reproducible on testnet**

- [x] This issue is only present on Devnet and cannot be reproduced on the alpha-sepolia testnet (check the box if true).

**To Reproduce**

1. Starting anvil with a block time and not instant mining with a small block time.
2. Sending some messages to L2.
3. Start starknet-devnet like so it can fetch several block with messages at one.
4. Verify the message nonce that should be (sometime due to the `HashMap`) in invalid order.

My mistake was to not test several blocks in the test code, and we're mostly having one block at a time, so no problem in this scenario.

**Devnet version**

- I am using Devnet version: main
- [x] This happens with a dockerized Devnet (check the box if true).
- This does not appear on the following Devnet version:

**System specifications**

- OS: MacOS 14.6.1

### PR DESCRIPTION
**Pull request by FabijanC**
## Usage related changes

- Close #586 

## Development related changes

- Use `BTreeMap` instead of `HashMap`
- Simplify code to reduce line count in `ethereum.rs`
- Add `spawn_with_additional_args` to `BackgroundAnvil`.
- Fix wrong messages in `BackgroundAnvil` method errors.
- Add utility method `felt_to_u256`.
- Minor refactoring in test `can_interact_with_l1` that improves readability.

## Checklist:

- [x] Checked out the [contribution guidelines](CONTRIBUTING.md)
- [x] Applied formatting - `./scripts/format.sh`
- [x] No linter errors - `./scripts/clippy_check.sh`
- [x] No unused dependencies - `./scripts/check_unused_deps.sh`
- [x] No spelling errors - `./scripts/check_spelling.sh`
- [x] Performed code self-review
- [x] Rebased to the latest commit of the target branch (or merged it into my branch)
    -   Once you make the PR reviewable, please avoid force-pushing
- [x] Updated the docs if needed - `./website/README.md`
- [x] Linked the [issues](https://github.com/0xSpaceShard/starknet-devnet-rs/issues) resolvable by this PR - [linking info](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)
- [x] Updated the tests if needed; all passing - [execution info](https://github.com/0xSpaceShard/starknet-devnet-rs/blob/main/.github/CONTRIBUTING.md#test-execution)

### CODE REVIEW

File: crates/starknet-devnet/tests/common/background_anvil.rs [hunk 1]
```diff
@@ -116,13 +121,13 @@ impl BackgroundAnvil {
...
                 TestError::EthersError(format!(
-                    "tx for deposit l1l2 contract on ethereum failed: {e}"
+                    "tx for withdraw l1l2 contract on ethereum failed: {e}"
                 ))
             })?
             .await
             .map_err(|e| {
                 TestError::EthersError(format!(
-                    "tx for deposit l1l2 contract on ethereum no receipt: {e}"
+                    "tx for withdraw l1l2 contract on ethereum has no receipt: {e}"
```
Comment (ivpavici) [inline crates/starknet-devnet/tests/common/background_anvil.rs:130]:
```suggestion
                    "tx for withdraw l1-l2 contract on ethereum has no receipt: {e}"
```

File: crates/starknet-devnet/tests/common/background_anvil.rs [hunk 2]
```diff
@@ -155,7 +160,7 @@ impl BackgroundAnvil {
             .await
             .map_err(|e| {
                 TestError::EthersError(format!(
-                    "tx for deposit l1l2 contract on ethereum no receipt: {e}"
+                    "tx for deposit l1l2 contract on ethereum has no receipt: {e}"
```
Comment (ivpavici) [inline crates/starknet-devnet/tests/common/background_anvil.rs:163]:
```suggestion
                    "tx for deposit l1-l2 contract on ethereum has no receipt: {e}"
```

File: crates/starknet-devnet/tests/common/utils.rs [hunk 3]
```diff
@@ -339,6 +340,11 @@ where
     }
 }
 
+/// Shall succeed if `f` fits into u128
```
Comment (ivpavici) [inline crates/starknet-devnet/tests/common/utils.rs:343]: Edwin in our ❤️

Comment (FabijanC) [reply crates/starknet-devnet/tests/common/utils.rs:343]: Actually this comment shalln't be here. It was applicable when I was using `u128` for conversion, but I changed that part in the meantime, so good that you left this comment.

File: crates/starknet-devnet/tests/common/background_anvil.rs (hunk 1 above)
Comment (FabijanC) [reply crates/starknet-devnet/tests/common/background_anvil.rs]: Addressed in https://github.com/0xSpaceShard/starknet-devnet-rs/pull/597/commits/b5040ca8266c91e5b4ab06979c8da4c318f40773

File: crates/starknet-devnet/tests/common/background_anvil.rs (hunk 2 above)
Comment (FabijanC) [reply crates/starknet-devnet/tests/common/background_anvil.rs:163]: Addressed in https://github.com/0xSpaceShard/starknet-devnet-rs/pull/597/commits/b5040ca8266c91e5b4ab06979c8da4c318f40773

File: crates/starknet-devnet/tests/common/utils.rs (hunk 3 above)
Comment (FabijanC) [reply crates/starknet-devnet/tests/common/utils.rs]: Removed in https://github.com/0xSpaceShard/starknet-devnet-rs/pull/597/commits/361f70f8fff139eb0a0e20087f3d4ff4eabe5868

**Review by ivpavici — APPROVED** [review]

**Pull request merged**